### PR TITLE
Use snapshots for hint arguments

### LINKED ISSUE #1846: Use array snapshots for cheatcode datatypes
To prevent errors with moved structures we should utilise snapshot mechanism for cheatcodes datatypes.

We should alter datatypes to the form
```
#[derive(Drop, Clone)]
struct PreparedContract {
    contract_address: felt252,
    class_hash: felt252,
    constructor_calldata: @Array::<felt252>,
}
```

Then, such code (I believe) should work
```
    let mut constructor_calldata = ArrayTrait::new();
    constructor_calldata.append(1);
    constructor_calldata.append(2);
    let class_hash = declare('with_ctor').unwrap();
    assert(class_hash != 0, 'declared class_hash != 0');

    let prepare_result = prepare(class_hash, constructor_calldata).unwrap();

    assert(prepare_result.contract_address != 0, 'prepared contract_address != 0');
    assert(prepare_result.class_hash != 0, 'prepared class_hash != 0');
    // TODO (1717): check the length of the array, this produces: error: Variable was previously moved
    assert(prepare_result.contract_address.len() == 2_u32, 'constructor_calldata size == 2');

```

### PR DESCRIPTION
**Pull request by cptartur**
Closes #1846 

Related cairo PR https://github.com/software-mansion-labs/cairo/pull/70

### Checklist
- [x] Relevant issue is linked
- [x] Docs updated/issue for docs created
- [x] Added relevant tests

### CODE REVIEW

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: tests/integration/cairo1/hint_locals/deploy/deploy_test.cairo [hunk 1]
```diff
@@ -70,19 +54,13 @@ fn test_deploy_with_storage() {
...
-    // TODO (1717): check the length of the array, this produces: error: Variable was previously moved
-    // assert(prepare_result.constructor_calldata.len() == 2_u32, 'constructor_calldata size == 2');
-
-    let prepared_contract = PreparedContract {
-        contract_address: prepare_result.contract_address,
-        class_hash: prepare_result.class_hash,
-        constructor_calldata: prepare_result.constructor_calldata
-    };
-    let deployed_contract_address = deploy(prepared_contract).unwrap();
+    assert(prepare_result.constructor_calldata.len() == 0_u32, 'constructor_calldata size == 0');
```
Comment (karol-bisztyga) [inline tests/integration/cairo1/hint_locals/deploy/deploy_test.cairo:61]:
```suggestion
    assert(prepare_result.constructor_calldata.len() == 2_u32, 'constructor_calldata size == 2');
    assert(*prepare_result.constructor_calldata.at(0_usize) == 1, 'constructor_calldata contents');
    assert(*prepare_result.constructor_calldata.at(1_usize) == 2, 'constructor_calldata contents');
```
nit, we could check this

File: tests/integration/cairo1/hint_locals/deploy/deploy_test.cairo [hunk 2]
```diff
@@ -135,14 +103,9 @@ fn test_deploy_with_ctor_obsolete_calldata() {
...
-    let prepare_result = prepare(class_hash, constructor_calldata).unwrap();
+    let prepare_result = prepare(class_hash, @constructor_calldata).unwrap();
 
-    let prepared_contract = PreparedContract {
-        contract_address: prepare_result.contract_address,
-        class_hash: prepare_result.class_hash,
-        constructor_calldata: prepare_result.constructor_calldata
-    };
-    match deploy(prepared_contract) {
+    match deploy(prepare_result) {
```
Comment (karol-bisztyga) [inline tests/integration/cairo1/hint_locals/deploy/deploy_test.cairo:108]: maybe we could add `test_deploy_doesnt_move_calldata` test case at the end of this test suite?

File: website/docs/tutorials/08-cairo-1-support/05-cheatcodes/README.md [hunk 3]
```diff
@@ -5,6 +5,48 @@ additional functions called cheatcodes that let you do modify the state beyond w
 
 Cheatcodes do not require any specific imports or `use` declarations.
 
+## Error handling in cheatcodes
+
+All cheatcodes return an `Result` enum:
```
Comment (karol-bisztyga) [inline website/docs/tutorials/08-cairo-1-support/05-cheatcodes/README.md:10]:
```suggestion
All cheatcodes return a `Result` enum:
```

File: website/docs/tutorials/08-cairo-1-support/05-cheatcodes/README.md [hunk 4]
```diff
@@ -5,6 +5,48 @@ additional functions called cheatcodes that let you do modify the state beyond w
...
+
+```cairo title="Result"
+enum Result<T, felt252> {
+    Ok: T,
+    Err: felt252,
+}
+```
```
[the remark this answers is not in the document]
Comment (cptartur) [reply website/docs/tutorials/08-cairo-1-support/05-cheatcodes/README.md:17]: I'll update the details. Problem is this information was duplicated on multiple pages. It makes it harder to maintain/will cause inconsistencies in the future.

Perhaps we can link to error handling from cheatcode pages?

File: website/docs/tutorials/08-cairo-1-support/05-cheatcodes/README.md [hunk 5]
```diff
@@ -5,6 +5,48 @@ additional functions called cheatcodes that let you do modify the state beyond w
...
+
+On successful cheatcode execution, `Ok` is returned - an exact type of `Ok` depends on the cheatcode.
```
Comment (karol-bisztyga) [inline website/docs/tutorials/08-cairo-1-support/05-cheatcodes/README.md:19]:
```suggestion
On successful cheatcode execution, a positive result is returned (`OK`) - its exact type depends on the cheatcode.
```

> an exact type of `Ok`

this sounds weird TBH.

File: website/docs/tutorials/08-cairo-1-support/05-cheatcodes/README.md [hunk 6]
```diff
@@ -5,6 +5,48 @@ additional functions called cheatcodes that let you do modify the state beyond w
...
+
+On failure, `Err` is returned,
+containing [short string](https://github.com/starkware-libs/cairo/blob/26188d4d3271c327fbbfd09f82c4acc99cb281f5/docs/reference/src/components/cairo/modules/language_constructs/pages/literal-expressions.adoc#short-string-literals)
+with encoded error message.
+
+### Failing test on cheatcode error
+
+Simplest handling of `Result` is to `unwrap()` them. It either returns a cheatcode success value or causes the test to
+fail entirely:
```
Comment (karol-bisztyga) [inline website/docs/tutorials/08-cairo-1-support/05-cheatcodes/README.md:28]:
```suggestion
The simplest handling of `Result` is to `unwrap()` them. It either returns a cheatcode success value or causes the test to
fail entirely:
```

Maybe we could also mention `expect` but it probably is not necessary.

File: website/docs/tutorials/08-cairo-1-support/05-cheatcodes/README.md [hunk 7]
```diff
@@ -5,6 +5,48 @@ additional functions called cheatcodes that let you do modify the state beyond w
...
+
+```cairo title="Simple handling"
+use result::ResultTrait;
+
+let prepared_contract = prepare(class_hash, @calldata).unwrap();
+```
+
+### Matching errors
+
+`Result` can be used inside `match` statement, to handle both `Ok` and `Err` types.
```
Comment (karol-bisztyga) [inline website/docs/tutorials/08-cairo-1-support/05-cheatcodes/README.md:38]:
```suggestion
`Result` can be used inside the `match` statement, to handle both `Ok` and `Err` types.
```

File: tests/integration/cairo1/hint_locals/deploy/deploy_test.cairo (hunk 2 above)
Comment (cptartur) [reply tests/integration/cairo1/hint_locals/deploy/deploy_test.cairo:108]: We should

**Review by Arcticae — APPROVED** [review]

File: tests/integration/cairo1/hint_locals/deploy/deploy_test.cairo (hunk 1 above)
Comment (karol-bisztyga) [reply tests/integration/cairo1/hint_locals/deploy/deploy_test.cairo:61]: I'm sorry I added this comment in a wrong place, it should've been added above 🤦

**Review by karol-bisztyga — APPROVED** [review]

**Comment by cptartur** [comment]
Not merging this before #1892 because of possible conflicts.

**Comment by MaksymilianDemitraszek** [comment]
@karol-bisztyga We will need to sync it with nightly compiler branch

**Comment by karol-bisztyga** [comment]
> @karol-bisztyga We will need to sync it with nightly compiler branch

I added changes there.

**Pull request merged**